operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1673 @@
+
+	SvcResource resource.Resource[*core_v1.Service]
+	SvcSynced   bool
+
+	PoolStore    PoolStore
+	RangesStore  RangesStore
+	ServiceStore ServiceStore
+
+	LBClasses   []string
+	IPv4Enabled bool
+	IPv6Enabled bool
+
+	// Only used during testing.
+	InitDone LBIPAMInitDone
+}
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	//
+	errChan := make(chan error, 2)
+
+	poolChan := stream.ToChannel[resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]](ctx, errChan, ipam.PoolResource)
+	svcChan := stream.ToChannel[resource.Event[*core_v1.Service]](ctx, errChan, ipam.SvcResource)
+
+	ipam.init(ctx, poolChan, svcChan)
+	if ipam.InitDone != nil {
+		close(ipam.InitDone)
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are closed to avoid blocking
+	// the resources in case they are re-used after we shutdown.
+	for poolChan != nil && svcChan != nil {
+		select {
+		case event, ok := <-poolChan:
+			if !ok {
+				poolChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.poolOnUpsert(ctx),
+				ipam.poolOnDelete(ctx),
+			)
+		case event, ok := <-svcChan:
+			if !ok {
+				svcChan = nil
+				continue
+			}
+
+			event.Handle(
+				nil,
+				ipam.svcOnUpsert(ctx),
+				ipam.svcOnDelete(ctx),
+			)
+		case err := <-errChan:
+			ipam.Logger.WithError(err).Error("Resource error")
+		}
+	}
+}
+
+func (ipam *LBIPAM) init(
+	ctx context.Context,
+	poolChan <-chan resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool],
+	svcChan <-chan resource.Event[*core_v1.Service],
+) {
+	// First sync all pools
+	poolSync := false
+	for !poolSync {
+		select {
+		case event := <-poolChan:
+			event.Handle(
+				func(s resource.Store[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]) error {
+					ipam.settleConflicts(ctx)
+					poolSync = true
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					ipam.handleNewPool(ctx, clbi)
+					return nil
+				},
+				func(k resource.Key, clbi *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+					ipam.Logger.Error("Got a Pool deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+
+	// Then sync all services
+	svcSync := false
+	for !svcSync {
+		select {
+		case event := <-svcChan:
+			event.Handle(
+				func(s resource.Store[*core_v1.Service]) error {
+					ipam.satisfyServices(ctx)
+					ipam.updateAllPoolCounts(ctx)
+					svcSync = true
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					ipam.handleUpsertService(ctx, svc)
+					return nil
+				},
+				func(k resource.Key, svc *core_v1.Service) error {
+					ipam.Logger.Error("Got a Service deleted event while syncing")
+					return nil
+				},
+			)
+		}
+	}
+}
+
+func (ipam *LBIPAM) poolOnUpsert(ctx context.Context) func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
+	return func(k resource.Key, pool *cilium_api_v2alpha1.CiliumLoadBalancerIPPool) error {
```

Comment:
I would keep the ctx in `*LBIPAM` so we don't need the closure. Or alternatively put these as methods in a struct that holds both ipam and ctx.